rfcs/0009-abstraction-layer-for-mobile-error-tracking.md

```diff
@@ -0,0 +1,296 @@
+
+Sentry.init({
+  dsn: // Sentry's DSN,
+});
+
+const errorTracking = createErrorTracking(Sentry, { enabled: true });
+
+const App = () => (
+  <ErrorTrackingProvider value={errorTracking}>
+    {/* ... */}
+  </ErrorTrackingProvider>
+);
+```
+
+## Usage examples
+
+Suppose there is a `Button.js` component:
+
+```js
+import React, { useCallback } from "react";
+import { TouchableOpacity, Text } from "react-native";
+import { useErrorTracking } from "../hooks/error-tracking";
+
+const errorTracking = useErrorTracking();
+
+const aFunctionThatMightFail = () => {
+  throw new Error("Failed");
+};
+
+const Button = () => {
+  const onPress = useCallback(() => {
+    try {
+      aFunctionThatMightFail();
+    } catch (err) {
+      errorTracking.captureError(err);
+    }
+  }, []);
+
+  return (
+    <TouchableOpacity onPress={onPress}>
+      <Text>Press Here</Text>
+    </TouchableOpacity>
+  );
+};
+```
+
+Note that it's possible to enable or disable the error tracking as you wish, so it would be possible to:
+
+```js
+import React, { useState, useCallback } from "react";
+import { View, ouchableOpacity, Text } from "react-native";
+import createErrorTracking from "@moxy/react-native-sentry";
+import { useErrorTracking } from "../hooks/error-tracking";
+
+const ToggleErrorTracking = () => {
+  const [enabled, setEnabled] = useState(true);
+
+  const errorTracking = useErrorTracking();
+
+  const onPress = useCallback(() => {
+    setEnabled((state) => !state);
+
+    errorTracking.enable(enabled);
+  }, [setEnabled, errorTracking]);
+
+  return (
+    <View>
+      <TouchableOpacity onPress={onPress}>
+        <Text>Toggle Error Tracking</Text>
+      </TouchableOpacity>
+
+      <Text>Error tracking is {enabled ? "enabled" : "disabled"}</Text>
+    </View>
+  );
+};
+```
+
+## SDK initialization
+
+In my opinion the initialization should not be abstracted, since Sentry and Firebase Crashlytics have quite differente ways of being initialized.
+
+### Sentry
+
+```js
+import * as Sentry from '@sentry/react-native';
+
+Sentry.init({
+    dsn: // Sentry's DSN,
+});
+```
+
+### Firebase Crashlytics
+
+```js
+import crashlytics from "@react-native-firebase/crashlytics";
+
+// you can already use it inside the file
+```
+
+# Drawbacks
+
+The only drawback I can see is the initial effort to design the solution, since the maintenance should be really simple.
+
+# Alternatives
+
+The alternative would be using the API provided directly from the error tracking solutions. However, this is not a good idea as I've already commented about in the "Motivation" section.
+
+# Adoption strategy
+
+To integrate those modules `@moxy/react-native-{solution}`, where `solution` would be the name of the chosen error tracking solution (it could be `sentry` for instance), one should start by installing it:
+
+```sh
+npm i @sentry/react-native @moxy/react-native-sentry
+
+# yarn add @sentry/react-native @moxy/react-native-sentry
```

Comment:
Improved the text a bit:

"New and existing apps are required to install the SDK of the chosen error tracking solution and the respective module which implements the abstraction for such SDK:

```sh
npm i @sentry/react-native @moxy/react-native-{solution}

# yarn add @sentry/react-native @moxy/react-native-{solution}
```

where `solution` would be the name of the solution to use, such as `crashlytics` or `sentry`.
Existing apps would also have to be refactored to start using the API specified here."

Reply:
👍